05-Mining-Protocol.md

```diff
@@ -1,101 +1,180 @@
+
+This process results in a Custom Job, where the template was was not unilaterally imposed by the Pool.
+
+## 5.2 Connections
+
+A new **Connection** is created as soon as:
+- the TCP communication is established between client and server.
+- the handshake process for Noise encryption has been completed.
+
+A new Connection under the Mining Protocol is established via the `SetupConnection` message, where the `protocol` message field is set as `0`.
+
+One Connection under the Mining Protocol can encapsulate multiple [Channels](#53-channels).
 
-The other portion of the block header that’s used to define the full search space is the Merkle root hash of all transactions in the block, projected to the last variable field in the block header:
+## 5.3 Channels
 
-- Merkle root, deterministically computed from:
-  - Coinbase transaction: typically 4-8 bytes, possibly much more.
-  - Transaction set: practically unbounded space.
-    All roles in Stratum v2 MUST NOT use transaction selection/ordering for additional hash space extension.
-    This stems both from the concept that miners/pools should be able to choose their transaction set freely without any interference with the protocol, and also to enable future protocol modifications to Bitcoin.
-    In other words, any rules imposed on transaction selection/ordering by miners not described in the rest of this document may result in invalid work/blocks.
+The Mining Protocol is designed such that downstream mining devices or proxies open **Channels** with upstream stratum nodes within established connections.
+The upstream stratum endpoints could be actual work-providing servers or proxies that pass the messages further upstream.
 
-Mining servers MUST assign a unique subset of the search space to each connection/channel (and therefore each mining device) frequently and rapidly enough so that the mining devices are not running out of search space.
-Unique jobs can be generated regularly by:
+Each Channel identifies a dedicated mining session associated with an authorized user.
+Upstream stratum nodes accept work submissions and specify a mining target on a per-channel basis.
+
+There can theoretically be up to `2^32` open Channels within one Connection.
+
+All channels are independent of each other, but share some messages broadcast from the server for higher efficiency (e.g. information about a new `prev_hash`).
+Each channel is identified by its `channel_id` (`U32`), which is consistent throughout the whole life of the Connection.
+
+A proxy can either transparently allow its clients to open separate Channels with the server (preferred behavior), or aggregate open connections from downstream devices into its own open channel with the server and translate the messages accordingly (present mainly for allowing v1 proxies).
+Both options have some practical use cases.
+In either case, proxies SHOULD aggregate clients' Channels into a smaller number of Connections.
```

Comment:
None of the SV1 miners are going to have a standard channel since they require a non-homogeneous setup. How does aggregation work in that case? At the same time, I have this question: each of these miners (clients) connects to the translator proxy, and each connection is a TCP connection. So how are we achieving multiplexing? My understanding was that with a single TCP connection, we could manage around 2^32 channels, but these channels should point to the same endpoint and starting point. Each of these miners connects to the same endpoint (the translator proxy), but they each have different ports. Each server can only handle a limited number of TCP connections, which is nowhere near 2^32—more like on the order of 10^5. So, the aggregation doesn’t happen at the level of the translator proxy and the miner, but rather between the translator proxy and another upstream connection, since they will have a single TCP connection through which we can have 2^32 channels for each downstream miner node. However, this raises a concern: the translator proxy becomes a single point of failure, so the provider needs to implement scaling strategies to address that.

Reply:
> None of the SV1 miners are going to have a standard channel since they require a non-homogeneous setup. How does aggregation work in that case?

tProxy makes sure that each SV1 miner is assigned a unique `extranonce_prefix`

Reply:
> My understanding was that with a single TCP connection, we could manage around 2^32 channels, but these channels should point to the same endpoint and starting point. Each of these miners connects to the same endpoint (the translator proxy), but they each have different ports. Each server can only handle a limited number of TCP connections, which is nowhere near 2^32—more like on the order of 10^5. So, the aggregation doesn’t happen at the level of the translator proxy and the miner, but rather between the translator proxy and another upstream connection, since they will have a single TCP connection through which we can have 2^32 channels for each downstream miner node. However, this raises a concern: the translator proxy becomes a single point of failure, so the provider needs to implement scaling strategies to address that.

I see the practical limitation you are pointing out here. However, I think this is easily solvable by stacking multiple proxy layers.

This limitation only manifests if we try to connect more than 10^5 machines into one single proxy. But if they are distributed across multiple proxies, there's no issue.

And if we imagine there's one "funnel" proxy aggregating all of them, and then forwarding that into some upstream layer (maybe a pool or JDC), this upstream layer would still be able to have up to 2^32 channels in a single connection.

Here's some comments about the practical aspects of this:
- https://github.com/plebhash/sv2-spec/blob/75e7c8c42a64b151d540ee1fd9f32d590ea37f26/02-Design-Goals.md?plain=1#L54
- https://github.com/plebhash/sv2-spec/blob/75e7c8c42a64b151d540ee1fd9f32d590ea37f26/10-Discussion.md?plain=1#L23